File: Cisco/11MaximumCandiesYouCanGetFromBoxes.py
from collections import deque


class Solution(object):
    def maxCandies(self, status, candies, keys, containedBoxes, initialBoxes):
        """
        :type status: List[int]
        :type candies: List[int]
        :type keys: List[List[int]]
        :type containedBoxes: List[List[int]]
        :type initialBoxes: List[int]
        :rtype: int
        """
        ready_boxes = deque()
        available_boxes = set()
        extra_keys = set()
        candies_collected = 0

        for box in initialBoxes:
            if status[box] == 1:
                ready_boxes.append(box)
            else:
                available_boxes.add(box)

        while ready_boxes:
            box = ready_boxes.pop()
            candy, new_boxes, new_keys = candies[box], containedBoxes[box], keys[box]

            candies_collected += candy
            extra_keys.update(new_keys)

            # For each new box:
            # if open => processing queue;
            # if closed => store for later;
            for new_box in new_boxes:
                if status[new_box] == 1:
                    ready_boxes.append(new_box)
                else:
                    available_boxes.add(new_box)

            newly_opened_boxes = extra_keys & available_boxes
            for newly_opened_box in newly_opened_boxes:
                ready_boxes.append(newly_opened_box)

            available_boxes -= newly_opened_boxes
            extra_keys -= newly_opened_boxes

        return candies_collected

File: Cisco/LinkedInList/C++/11MaximumCandiesYouCanGetFromBoxes.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution
{
public:
    int maxCandies(vector<int> &status, vector<int> &candies,
                   vector<vector<int>> &keys, vector<vector<int>> &containedBoxes, vector<int> &initialBoxes)
    {
        int NUM_BOXES = status.size();
        queue<int> q;
        vector<bool> reachableClosed(NUM_BOXES, false);

        for (const auto &box : initialBoxes)
            if (status[box])
                q.push(box);
            else
                reachableClosed[box] = true;

        int candiesCollected = 0;
        while (!q.empty())
        {
            int box = q.front();
            q.pop();
            candiesCollected += candies[box];

            for (const auto &key : keys[box])
            {
                if (!status[key] && reachableClosed[key])
                    q.push(key);
                status[key] = 1; // Reachable or not - unlocking it (this is neat trick) :)
            }

            for (const auto &containedBox : containedBoxes[box])
                if (status[containedBox])
                    q.push(containedBox);
                else
                    reachableClosed[containedBox] = true;
        }

        return candiesCollected;
    }
};